lib/phoenix/pubsub/local.ex

```diff
@@ -0,0 +1,212 @@
+
+  ## Examples
+
+      iex> PubSub.Local.subscribers(:local_server, "foo")
+      #HashSet<[]>
+
+  """
+  def subscribers(local_server, topic) do
+    GenServer.call(local_server, {:subscribers, topic})
+  end
+
+  @doc false
+  def list(local_server) do
+    GenServer.call(local_server, :list)
+  end
+
+  @doc false
+  def subscription(local_server, pid) do
+    GenServer.call(local_server, {:subscription, pid})
+  end
+
+  @doc false
+  def init(_) do
+    Process.flag(:trap_exit, true)
+    {:ok, %{topics: HashDict.new, pids: HashDict.new}}
+  end
+
+  def handle_call(:list, _from, state) do
+    {:reply, Dict.keys(state.topics), state}
+  end
+
+  def handle_call({:subscription, pid}, _from, state) do
+    case HashDict.fetch(state.pids, pid) do
+      {:ok, {_ref, topics}} -> {:reply, topics, state}
+      :error                -> {:reply, :no_subscription, state}
+    end
+  end
+
+  def handle_call({:subscribers, topic}, _from, state) do
+    {:reply, HashDict.get(state.topics, topic, HashSet.new), state}
+  end
+
+  def handle_call({:subscribe, pid, topic}, _from, state) do
+    {:reply, :ok, put_subscription(state, pid, topic)}
+  end
+
+  def handle_call({:unsubscribe, pid, topic}, _from, state) do
+    {:reply, :ok, drop_subscription(state, pid, topic)}
+  end
+
+  def handle_call({:broadcast, from_pid, topic, msg}, _from, state) do
+    case HashDict.fetch(state.topics, topic) do
+      {:ok, pids} ->
+        Enum.each(pids, fn
+          pid when pid != from_pid -> send(pid, msg)
+          _ -> :ok
+        end)
+        {:reply, :ok, state}
+
+      :error ->
+        {:reply, :no_topic, state}
+    end
+  end
+
+  def handle_info({:DOWN, ref, _type, pid, _info}, state) do
+    {:noreply, drop_subscriber(state, pid, ref)}
+  end
+
+  def terminate(_reason, _state) do
+    # TODO notify linked (?) subscribers?
+    :ok
+  end
+
+  defp put_subscription(state, pid, topic) do
+    subscription = case HashDict.fetch(state.pids, pid) do
+      {:ok, {ref, topics}} ->
+        {ref, HashSet.put(topics, topic)}
+      :error ->
+        {Process.monitor(pid), HashSet.put(HashSet.new, topic)}
+    end
+
+    topic_pids = case HashDict.fetch(state.topics, topic) do
+      {:ok, pids} -> HashSet.put(pids, pid)
+      :error      -> HashSet.put(HashSet.new, pid)
+    end
+
+    %{state | topics: HashDict.put(state.topics, topic, topic_pids),
+              pids: HashDict.put(state.pids, pid, subscription)}
+  end
+
+  defp drop_subscription(state, pid, topic) do
+    case HashDict.fetch(state.topics, topic) do
+      :error      -> state
+      {:ok, topic_pids} ->
+        case HashDict.fetch(state.pids, pid) do
+          :error      -> state
+          {:ok, {ref, subd_topics}} ->
+            topic_pids  = HashSet.delete(topic_pids, pid)
+            subd_topics = HashSet.delete(subd_topics, topic)
+
+            cond do
+              Enum.any?(topic_pids) && Enum.any?(subd_topics) ->
```

Comment:
Actually, it seems if you handle both cases separately, you can get rid of the duplication:

```elixir
topics = 
  if Enum.any?(topic_pids) do
    HashDict.put(state.topics, topic, topic_pids)
  else
    HashDict.delete(state.topics, topic)
  end

pids =
  if Enum.any?(subd_topics) do
    HashDict.put(state.pids, pid, {ref, subd_topics})
  else
    HashDict.delete(state.pids, pid)
  end
```